test/ZNSSubRegistrar.test.ts

```diff
@@ -122,22 +136,423 @@ describe("ZNSSubRegistrar", () => {
+
+      await zns.subRegistrar.connect(lvl3SubOwner).registerMultipleSubdomains(registrations);
+
+      const logs = await getDomainRegisteredEvents({
+        zns,
+        registrant: lvl3SubOwner.address,
+      });
+
+      for (let i = 0; i < domainLevels; i++) {
+        const subdomain = registrations[i];
+
+        // "DomainRegistered" event log
+        const args = logs[i].args;
+
+        i > 0 ?
+          expect(args[0]).to.eq(parentHashes[i - 1]) :
+          expect(args[0]).to.eq(rootHash);              // parentHash
+        expect(args[1]).to.eq(parentHashes[i]);         // domainHash current
+        expect(args[2]).to.eq(subdomain.label);         // label
+        expect(args[4]).to.eq(subdomain.tokenURI);      // tokenURI
+        expect(args[5]).to.eq(lvl3SubOwner.address);    // registrant
+        expect(args[6]).to.eq(subdomain.domainAddress); // domainAddress
+      }
+    });
+
+    it("Should revert when register the same domain twice using #registerMultipleSubdomains", async () => {
+      const subdomainObj = {
+        parentHash: rootHash,
+        label: "subdomain1",
+        domainAddress: admin.address,
+        tokenURI: "0://tokenURI",
+        distributionConfig: {
+          pricerContract: await zns.curvePricer.getAddress(),
+          paymentType: PaymentType.STAKE,
+          accessType: AccessType.LOCKED,
+        },
+        paymentConfig: {
+          token: await zns.meowToken.getAddress(),
+          beneficiary: admin.address,
+        },
+      };
+
+      // Add allowance
+      await zns.meowToken.connect(lvl2SubOwner).approve(await zns.treasury.getAddress(), ethers.MaxUint256);
+
+      await expect(
+        zns.subRegistrar.connect(lvl2SubOwner).registerMultipleSubdomains([subdomainObj, subdomainObj])
+      ).to.be.revertedWithCustomError(zns.subRegistrar, "DomainAlreadyExists");
+    });
+
+    it("Should revert with #ZeroAddressPassed when 1st level subdomain has zerohash", async () => {
+      const subdomainObj = {
+        parentHash: ethers.ZeroHash,
+        label: "subdomainzeroaAddresspassed",
+        domainAddress: admin.address,
+        tokenURI: "0://tokenURI",
+        distributionConfig: {
+          pricerContract: await zns.curvePricer.getAddress(),
+          paymentType: PaymentType.STAKE,
+          accessType: AccessType.LOCKED,
+        },
+        paymentConfig: {
+          token: await zns.meowToken.getAddress(),
+          beneficiary: admin.address,
+        },
+      };
+
+      // Add allowance
+      await zns.meowToken.connect(lvl2SubOwner).approve(await zns.treasury.getAddress(), ethers.MaxUint256);
+
+      await expect(
+        zns.subRegistrar.connect(lvl2SubOwner).registerMultipleSubdomains([subdomainObj])
+      ).to.be.revertedWithCustomError(zns.subRegistrar, ZERO_PARENTHASH_ERR);
+    });
+
+    it("Should register a mix of nested and non-nested subdomains and validates hashes", async () => {
```

Comment:
regarding some discussions we've had recently, it may be best to organize this into multiple smaller tests that are within a `describe` suite. This way you avoid having a single ~230 line test for quite a few different scenarios and instead have a test specific for each one. This also helps because it is more specific as to what is failing when tests are separate. Currently if this test fails there could be many potential reasons